test/apps/test_hooks.c

```diff
@@ -0,0 +1,240 @@
+/**
+ * Copyright (c) NVIDIA CORPORATION & AFFILIATES, 2023. ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include <sys/mman.h>
+#include <sys/time.h>
+#include <dlfcn.h>
+#include <stdio.h>
+#include <unistd.h>
+#include <stdlib.h>
+#include <pthread.h>
+
+#include <ucm/api/ucm.h>
+
+
+#define _QUOTE(x) #x
+#define QUOTE(x)  _QUOTE(x)
+
+#define DEFAULT_THREAD_COUNT   16
+#define DEFAULT_TIMEOUT        10
+#define DEFAULT_THREAD_TIMEOUT 1
+
+#define log(fmt, ...) printf(fmt "\n", ##__VA_ARGS__)
+
+
+static void *mmap_call(size_t size)
+{
+    return mmap(NULL, size, PROT_READ | PROT_WRITE, MAP_PRIVATE, -1, 0);
+}
+
+struct context {
+    const char   *libname;
+    int          thread_count;
+    pthread_t    *thread;
+    volatile int stop;
+    uint64_t     thread_timeout_msec;
+    void         *dl_handle;
+};
+
+static ucs_status_t ctx_init(struct context *ctx, const char *filename,
+                             int thread_count, int thread_timeout)
+{
+    pthread_t *thread;
+
+    thread = calloc(thread_count, sizeof(*thread));
+    if (thread == NULL) {
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    ctx->dl_handle           = NULL;
+    ctx->libname             = filename;
+    ctx->thread_count        = thread_count;
+    ctx->thread              = thread;
+    ctx->stop                = 0;
+    ctx->thread_timeout_msec = thread_timeout * 1000;
+
+    return UCS_OK;
+}
+
+static ucs_status_t ctx_load_library(struct context *ctx)
+{
+    log("loading library to trigger hooks installation");
+
+    dlerror();
+    ctx->dl_handle = dlopen(ctx->libname, RTLD_NOW);
+    if (ctx->dl_handle == NULL) {
+        log("dlopen(filename=\"%s\", RTLD_NOW) failed: %s", ctx->libname,
+            dlerror());
+        return UCS_ERR_INVALID_PARAM;
+    }
+
+    return UCS_OK;
+}
+
+static uint64_t get_time_msec(void)
```

Comment:
maybe use ucs_get_time()?

Reply:
cannot use it as it brings dependency on libucs code that we are dlopen'ing.